File: Math/discrete_log.cpp
ll discrete_log(ll p, ll g, ll h){
	// baby step giant step
	// O(sqrt{p} log p)
	
	// find x in [0, P) such that
	// g^x == h (mod p)
	// g is primitive root of  p
	// p is prime
	
	map<ll, ll> m;
	
	ll b = ll(sqrt(p-1)) + 1;
		
	for (ll v = 0; v < b; v++){
		ll tmp = power(g, p-2, p);
		m.insert({mul(h, power(tmp, v, p), p), v});
	}
		
	ll ans = p;
	g = power(g, b, p);
	for (ll u = 0; u < b; u++){
		ll tmp = power(g, u, p);
		if (m.find(tmp) != m.end()){
			ans = u*b + m[tmp];
			break;
		}
	}
	
	return ans;
}

File: Math/discrete_log.py
def discrete_log(p, g, h):
    # return x in [0, p) s.t. g^x = h (mod p)
    from math import isqrt

    m = dict()
    b = isqrt(p-1)+1

    for v in range(b):
        tmp = pow(g, p-2, p)
        m[h*pow(g, v*(p-2), p)%p] = v
    
    ans = p
    g = pow(g, b, p)
    for u in range(b):
        tmp = pow(g, u, p)
        if tmp in m.keys():
            ans = u*b + m[tmp]
            break
    
    return ans
